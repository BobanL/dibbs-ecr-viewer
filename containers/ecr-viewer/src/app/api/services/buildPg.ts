import { Kysely } from "kysely";
import { Core } from "./core_types";
import { Extended } from "./extended_types";
import { dialect } from "./dialects/postgres";

export const pgConstructor = (schema: "core" | "extended") => {
    if (schema === "core") {
        return new Kysely<Core>(dialect);
    } else if (schema === "extended") {
        return new Kysely<Extended>(dialect);
    } else {
        throw new Error("Invalid schema type.");
    }
}