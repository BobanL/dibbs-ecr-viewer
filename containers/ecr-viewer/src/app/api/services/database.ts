// Kysely ORM Connection Client

import { Kysely } from "kysely";

import { dialect as postgres } from "./dialects/postgres";
import { dialect as sqlserver } from "./dialects/sqlserver";
import { Common } from "./types/common";

// Dialect to communicate with the database, interface to define its structure.

let db: unknown;

/**
 * Get the current database dialect
 * @returns string describing dialect
 */
export const dbDialect = () => {
  return process.env.METADATA_DATABASE_TYPE;
};

/**
 * Get the current database schema
 * @returns string describing schema
 */
export const dbSchema = () => {
  return process.env.METADATA_DATABASE_SCHEMA;
};

/**
 * Get the current database namespace (schema)
 * @returns string describing namespace
 */
export const dbNamespace = () => {
  // use a different schema in testing so seed data doesn't get wiped out
  return process.env.TEST_TYPE === "integration"
    ? "test_ev_schema"
    : "ecr_viewer";
};

/**
 * Get the database global.
 * @returns global db
 */
export const getDb = <T>() => {
  if (db) {
    return db as Kysely<T>;
  }

  const db_type = dbDialect();
  switch (db_type) {
    case "sqlserver":
      db = new Kysely(sqlserver);
      break;
    case "postgres":
      db = new Kysely(postgres);
      break;
    default:
      throw new Error(`unknown db type: ${db_type}`);
  }

  db = (db as Kysely<T>).withSchema(dbNamespace());

  return db as Kysely<T>;
};

/**
 * Performs a health check on the metadata database connection.
 * @returns The status of the metadata db connection or undefined if missing environment values.
 */
export const metadataDatabaseHealthCheck = async () => {
  if (!process.env.METADATA_DATABASE_TYPE) {
    return undefined;
  }
  try {
    await getDb<Common>()
      .connection()
      .execute(async (_db) => {});
    return "UP";
  } catch (error: unknown) {
    console.error(error);
    return "DOWN";
  }
};
