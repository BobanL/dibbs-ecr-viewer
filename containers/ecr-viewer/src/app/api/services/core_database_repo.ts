import { Kysely } from "kysely";

import {
  ECRConditions,
  NewECRConditions,
  ECRConditionsUpdate,
  ECRRuleSummaries,
  NewECRRuleSummaries,
  ECRRuleSummariesUpdate,
  ECR,
  NewECR,
  ECRUpdate,
} from "./core_types";
import { Core } from "./core_types";
import { db } from "./database";

const coredb = db as Kysely<Core>;

// ECR_DATA

/**
 * Finds an eICR by its ID
 * @async
 * @function findEcrById
 * @param id - thd ID of the eCR to be looked up
 * @returns an eICR object
 */
export async function findEcrById(id: string | null): Promise<ECR | undefined> {
  if (!id) {
    throw new Error("eICR ID is required.");
  }
  try {
    return await coredb
      .selectFrom("ecr_data")
      .where("eICR_ID", "=", id)
      .selectAll()
      .executeTakeFirst();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Finds an eICR by its criteria
 * @async
 * @function findEcr
 * @param criteria - the Partial<ECR> used to filter against
 * @returns an eICR object
 */
export async function findEcr(criteria: Partial<ECR> | null): Promise<ECR[]> {
  let query = coredb.selectFrom("ecr_data");

  if (!criteria) {
    throw new Error("eICR Criteria is required.");
  }

  for (const criterium of Object.keys(criteria) as (keyof ECR)[]) {
    if (criteria[criterium] !== undefined && criteria[criterium] !== null) {
      query = query.where(criterium, "=", criteria[criterium]);
    }
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eICR object
 * @async
 * @function createEcr
 * @param ecr - the NewECR to be persisted
 * @returns the created eICR object
 */
export async function createEcr(ecr: NewECR | null): Promise<ECR | undefined> {
  if (!ecr) {
    throw new Error("eICR Data is required.");
  }
  try {
    return await coredb
      .insertInto("ecr_data")
      .values(ecr)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eICR object
 * @async
 * @function updateEcr
 * @param eICR_ID - the eICR_ID of the eCR to be updated
 * @param updateWith - the ECRUpdate to be applied to the existing record
 * @returns the updated eICR object
 */
export async function updateEcr(
  eICR_ID: string | null,
  updateWith: ECRUpdate,
): Promise<void> {
  await coredb
    .updateTable("ecr_data")
    .set(updateWith)
    .where("eICR_ID", "=", eICR_ID)
    .execute();
}

/**
 * Deletes an eICR object
 * @async
 * @function deleteEcr
 * @param eICR_ID - the eICR_ID of the record to be deleted
 * @returns the deleted eICR object
 */
export async function deleteEcr(
  eICR_ID: string | null,
): Promise<ECR | undefined> {
  const ecr = await findEcrById(eICR_ID);

  if (!ecr) {
    return undefined;
  }

  await coredb.deleteFrom("ecr_data").where("eICR_ID", "=", eICR_ID).execute();

  return ecr;
}

// ECR_RR_CONDITIONS

/**
 * Finds an eCR condition by its ID
 * @async
 * @function findEcrConditionById
 * @param id - the UUID of the eCR condition being looked up
 * @returns an eCR condition object
 */
export async function findEcrConditionById(
  id: string,
): Promise<ECRConditions | undefined> {
  return await coredb
    .selectFrom("ecr_rr_conditions")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR condition by its criteria
 * @async
 * @function findEcrCondition
 * @param criteria - the Partial<ECRConditions> filtering criteria to be applied
 * @returns an eCR condition object
 */
export async function findEcrCondition(
  criteria: Partial<ECRConditions>,
): Promise<ECRConditions[]> {
  let query = coredb.selectFrom("ecr_rr_conditions");

  for (const criterium of Object.keys(criteria) as (keyof ECRConditions)[]) {
    if (criteria[criterium] !== undefined && criteria[criterium] !== null) {
      query = query.where(criterium, "=", criteria[criterium]);
    }
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR condition object
 * @async
 * @function createEcrCondition
 * @param condition - the NewECRConditions to be created
 * @returns the created eCR condition object
 */
export async function createEcrCondition(
  condition: NewECRConditions | null,
): Promise<ECRConditions | undefined> {
  if (!condition) {
    throw new Error("eICR Data is required.");
  }
  try {
    return await coredb
      .insertInto("ecr_rr_conditions")
      .values(condition)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR condition object
 * @async
 * @function updateEcrCondition
 * @param uuid - the UUID of the eCR Condition to be updated
 * @param updateWith - the ECRConditionsUpdate to be applied
 * @returns the updated eCR condition object
 */
export async function updateEcrCondition(
  uuid: string,
  updateWith: ECRConditionsUpdate,
): Promise<void> {
  await coredb
    .updateTable("ecr_rr_conditions")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR condition object
 * @async
 * @function deleteEcrCondition
 * @param uuid - thge UUID of the record to be deleted
 * @returns the deleted eCR condition object
 */
export async function deleteEcrCondition(
  uuid: string,
): Promise<ECRConditions | undefined> {
  const ecr = await findEcrConditionById(uuid);

  if (ecr) {
    await coredb
      .deleteFrom("ecr_rr_conditions")
      .where("uuid", "=", uuid)
      .execute();
  }

  return ecr;
}

// ECR_RR_RULE_SUMMARIES

/**
 * Finds an eCR rule summary by its ID
 * @async
 * @function findEcrRuleById
 * @param id - the ID of the eCR being looked up
 * @returns an eCR rule object
 */
export async function findEcrRuleById(
  id: string,
): Promise<ECRRuleSummaries | undefined> {
  return await coredb
    .selectFrom("ecr_rr_rule_summaries")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR rule summary by its criteria
 * @async
 * @function findEcrRule
 * @param criteria - the Partial<ECRRuleSummaries> filtering criteria
 * @returns an eCR rule object
 */
export async function findEcrRule(
  criteria: Partial<ECRRuleSummaries>,
): Promise<ECRRuleSummaries[]> {
  let query = coredb.selectFrom("ecr_rr_rule_summaries");

  for (const criterium of Object.keys(criteria) as (keyof ECRRuleSummaries)[]) {
    if (criteria[criterium] !== undefined && criteria[criterium] !== null) {
      query = query.where(criterium, "=", criteria[criterium]);
    }
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR rule summary object
 * @async
 * @param rule_summary - the NewECRRuleSummaries record to be created
 * @function createEcrRule
 * @returns the created eCR rule object
 */
export async function createEcrRule(
  rule_summary: NewECRRuleSummaries,
): Promise<ECRRuleSummaries | undefined> {
  if (!rule_summary) {
    throw new Error("eICR Data is required.");
  }
  try {
    return await coredb
      .insertInto("ecr_rr_rule_summaries")
      .values(rule_summary)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR rule summary object
 * @async
 * @function updateEcrRule
 * @param uuid - the UUID of the eCR rule to be updated
 * @param updateWith - the ECRRuleSummariesUpdate to be applied to the record
 * @returns the updated eCR rule object
 */
export async function updateEcrRule(
  uuid: string,
  updateWith: ECRRuleSummariesUpdate,
): Promise<void> {
  await coredb
    .updateTable("ecr_rr_rule_summaries")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR rule summary object
 * @async
 * @function deleteEcrRule
 * @param uuid - the UUID of the eCR rule to be deleted
 * @returns the deleted eCR rule object
 */
export async function deleteEcrRule(
  uuid: string,
): Promise<ECRRuleSummaries | undefined> {
  const rule_summary = await findEcrRuleById(uuid);

  if (rule_summary) {
    await coredb
      .deleteFrom("ecr_rr_rule_summaries")
      .where("uuid", "=", uuid)
      .execute();
  }

  return rule_summary;
}
