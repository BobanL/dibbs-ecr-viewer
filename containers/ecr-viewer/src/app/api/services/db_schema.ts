import { Kysely, sql } from "kysely";
import { db } from "./database";
import { Core } from "./core_types";
import { Extended } from "./extended_types";

/**
 * Builds the extended schema to a test database
 * @async
 * @function buildExtended
 */
export const buildExtended = async () => {
    await (db as Kysely<Extended>).schema
      .createTable("ecr_data")
      .addColumn("eICR_ID", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("set_id", "varchar(255)")
      .addColumn("fhir_reference_link", "varchar(255)")
      .addColumn("last_name", "varchar(255)", (cb) => cb.notNull())
      .addColumn("first_name", "varchar(255)", (cb) => cb.notNull())
      .addColumn("birth_date", "date", (cb) => cb.notNull())
      .addColumn("gender", "varchar(100)")
      .addColumn("birth_sex", "varchar(255)")
      .addColumn("gender_identity", "varchar(255)")
      .addColumn("race", "varchar(255)")
      .addColumn("ethnicity", "varchar(255)")
      .addColumn("latitude", "numeric")
      .addColumn("longitude", "numeric")
      .addColumn("homelessness_status", "varchar(255)")
      .addColumn("disabilities", "varchar(255)")
      .addColumn("tribal_affiliation", "varchar(255)")
      .addColumn("tribal_enrollment_status", "varchar(255)")
      .addColumn("current_job_title", "varchar(255)")
      .addColumn("current_job_industry", "varchar(255)")
      .addColumn("usual_occupation", "varchar(255)")
      .addColumn("usual_industry", "varchar(255)")
      .addColumn("preferred_language", "varchar(255)")
      .addColumn("pregnancy_status", "varchar(255)")
      .addColumn("rr_id", "varchar(255)")
      .addColumn("processing_status", "varchar(255)")
      .addColumn("eicr_version_number", "varchar(50)")
      .addColumn("authoring_date", "date")
      .addColumn("authoring_provider", "varchar(255)")
      .addColumn("provider_id", "varchar(255)")
      .addColumn("facility_id", "varchar(255)")
      .addColumn("facility_name", "varchar(255)")
      .addColumn("encounter_type", "varchar(255)")
      .addColumn("encounter_start_date", "date")
      .addColumn("encounter_end_date", "date")
      .addColumn("reason_for_visit", "text")
      .addColumn("active_problems", "text")
      .addColumn("date_created", "timestamptz", (cb) =>
        cb.notNull().defaultTo(sql`NOW()`),
      )
      .execute();
    await (db as Kysely<Extended>).schema
      .createTable("patient_address")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("use", "varchar(50)")
      .addColumn("type", "varchar(50)")
      .addColumn("text", "varchar(255)")
      .addColumn("line", "varchar(255)")
      .addColumn("city", "varchar(100)")
      .addColumn("district", "varchar(100)")
      .addColumn("state", "varchar(100)")
      .addColumn("postal_code", "varchar(20)")
      .addColumn("country", "varchar(100)")
      .addColumn("period_start", "date")
      .addColumn("period_end", "date")
      .addColumn("eICR_ID", "varchar(200)")
      .execute();
    await (db as Kysely<Extended>).schema
      .createTable("ecr_labs")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("eICR_ID", "varchar(200)")
      .addColumn("test_type", "varchar(255)")
      .addColumn("test_type_code", "varchar(255)")
      .addColumn("test_type_system", "varchar(255)")
      .addColumn("test_result_qualitative", "varchar(255)")
      .addColumn("test_result_quantitative", "numeric")
      .addColumn("test_result_units", "varchar(50)")
      .addColumn("test_result_code", "varchar(255)")
      .addColumn("test_result_code_display", "varchar(255)")
      .addColumn("test_result_code_system", "varchar(255)")
      .addColumn("test_result_interpretation", "varchar(255)")
      .addColumn("test_result_interpretation_code", "varchar(255)")
      .addColumn("test_result_interpretation_system", "varchar(255)")
      .addColumn("test_result_reference_range_low_value", "numeric")
      .addColumn("test_result_reference_range_low_units", "varchar(50)")
      .addColumn("test_result_reference_range_high_value", "numeric")
      .addColumn("test_result_reference_range_high_units", "varchar(50)")
      .addColumn("specimen_type", "varchar(255)")
      .addColumn("specimen_collection_date", "date")
      .addColumn("performing_lab", "varchar(255)")
      .execute();
    await (db as Kysely<Extended>).schema
      .createTable("ecr_rr_conditions")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("eICR_ID", "varchar(255)", (cb) => cb.notNull())
      .addColumn("condition", "varchar")
      .execute();
    await (db as Kysely<Extended>).schema
      .createTable("ecr_rr_rule_summaries")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("ecr_rr_conditions_id", "varchar(200)")
      .addColumn("rule_summary", "varchar")
      .execute();
}

/**
 * Drops the extended schema from a test database
 * @async
 * @function dropExtended
 */
export const dropExtended = async () => {
    await (db as Kysely<Extended>).schema.dropTable("ecr_data").execute();
    await (db as Kysely<Extended>).schema.dropTable("patient_address").execute();
    await (db as Kysely<Extended>).schema.dropTable("ecr_labs").execute();
    await (db as Kysely<Extended>).schema.dropTable("ecr_rr_conditions").execute();
    await (db as Kysely<Extended>).schema.dropTable("ecr_rr_rule_summaries").execute();
}

/**
 * Clears the extended schema tables on a test database
 * @async
 * @function clearExtended
 */
export const clearExtended = async () => {
    await (db as Kysely<Extended>).deleteFrom("ecr_data").execute();
    await (db as Kysely<Extended>).deleteFrom("patient_address").execute();
    await (db as Kysely<Extended>).deleteFrom("ecr_labs").execute();
    await (db as Kysely<Extended>).deleteFrom("ecr_rr_conditions").execute();
    await (db as Kysely<Extended>).deleteFrom("ecr_rr_rule_summaries").execute();
}

/**
 * Builds the core schema to a test database
 * @async
 * @function buildCore
 */
export const buildCore = async () => {
    await (db as Kysely<Core>).schema
      .createTable("ecr_data")
      .addColumn("eICR_ID", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("set_id", "varchar(255)")
      .addColumn("eicr_version_number", "varchar(50)")
      .addColumn("data_source", "varchar(2)", (cb) => cb.notNull()) // S3 or DB
      .addColumn("fhir_reference_link", "varchar(500)")
      .addColumn("patient_name_first", "varchar(100)", (cb) => cb.notNull())
      .addColumn("patient_name_last", "varchar(100)", (cb) => cb.notNull())
      .addColumn("patient_birth_date", "date", (cb) => cb.notNull())
      .addColumn("date_created", "timestamptz", (cb) =>
        cb.notNull().defaultTo(sql`NOW()`),
      )
      .addColumn("report_date", "date", (cb) => cb.notNull())
      .execute();
    await (db as Kysely<Core>).schema
      .createTable("ecr_rr_conditions")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("eICR_ID", "varchar(255)", (cb) => cb.notNull())
      .addColumn("condition", "varchar")
      .execute();
    await (db as Kysely<Core>).schema
      .createTable("ecr_rr_rule_summaries")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("ecr_rr_conditions_id", "varchar(200)")
      .addColumn("rule_summary", "varchar")
      .execute();
}

/**
 * Drops the core schema from a test database
 * @async
 * @function dropCore
 */
export const dropCore = async () => {
    await (db as Kysely<Core>).schema.dropTable("ecr_data").execute();
    await (db as Kysely<Core>).schema.dropTable("ecr_rr_conditions").execute();
    await (db as Kysely<Core>).schema.dropTable("ecr_rr_rule_summaries").execute();
}

/**
 * Clears the core schema tables on a test database
 * @async
 * @function clearCore
 */
export const clearCore = async () => {
    await (db as Kysely<Core>).deleteFrom("ecr_data").execute();
    await (db as Kysely<Core>).deleteFrom("ecr_rr_conditions").execute();
    await (db as Kysely<Core>).deleteFrom("ecr_rr_rule_summaries").execute();
}

/**
 * Builds the aliased schema to a test database
 * @async
 * @function buildCoreAlias
 */
export const buildCoreAlias = async () => {
    await db.schema
      .createTable("ecr_viewer.ecr_data")
      .addColumn("eicr_id", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("set_id", "varchar(255)")
      .addColumn("eicr_version_number", "varchar(50)")
      .addColumn("data_source", "varchar(2)") // S3 or DB
      .addColumn("fhir_reference_link", "varchar(500)")
      .addColumn("patient_name_first", "varchar(100)")
      .addColumn("patient_name_last", "varchar(100)")
      .addColumn("patient_birth_date", "date")
      .addColumn("date_created", "timestamptz", (cb) =>
        cb.notNull().defaultTo(sql`NOW()`),
      )
      .addColumn("report_date", "date")
      .execute();
    await db.schema
      .createTable("ecr_viewer.ecr_rr_conditions")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("eicr_id", "varchar(255)", (cb) => cb.notNull())
      .addColumn("condition", "varchar")
      .execute();
    await db.schema
      .createTable("ecr_viewer.ecr_rr_rule_summaries")
      .addColumn("uuid", "varchar(200)", (cb) => cb.primaryKey())
      .addColumn("ecr_rr_conditions_id", "varchar(200)")
      .addColumn("rule_summary", "varchar")
      .execute();
}

/**
 * Drops the aliased schema from a test database
 * @async
 * @function dropCoreAlias
 */
export const dropCoreAlias = async () => {
    await db.schema.dropTable("ecr_viewer.ecr_data").execute();
    await db.schema.dropTable("ecr_viewer.ecr_rr_conditions").execute();
    await db.schema.dropTable("ecr_viewer.ecr_rr_rule_summaries").execute();
}

/**
 * Clears the aliased schema tables on a test database
 * @async
 * @function clearCoreAlias
 */
export const clearCoreAlias = async () => {
    await (db as any).deleteFrom("ecr_viewer.ecr_data").execute();
    await (db as any).deleteFrom("ecr_viewer.ecr_rr_conditions").execute();
    await (db as any).deleteFrom("ecr_viewer.ecr_rr_rule_summaries").execute();
}