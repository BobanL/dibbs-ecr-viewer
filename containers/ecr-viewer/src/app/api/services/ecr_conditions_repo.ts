import { db } from './database'
import { ECRConditions, NewECRConditions, ECRConditionsUpdate } from './types'

export async function findEcrConditionById(id: string) {
  return await db.selectFrom('ecr_rr_conditions')
    .where('uuid', '=', id)
    .selectAll()
    .executeTakeFirst()
}

export async function findEcrCondition(criteria: Partial<ECRConditions>) {
  let query = db.selectFrom('ecr_rr_conditions')

  if (criteria.uuid) {
    query = query.where('uuid', '=', criteria.uuid)
  }

  if (criteria.eICR_ID) {
    query = query.where('eICR_ID', '=', criteria.eICR_ID)
  }

  if (criteria.condition) {
    query = query.where('condition', '=', criteria.condition)
  }

  return await query.selectAll().execute()
}

export async function createEcrCondition(ecr: NewECRConditions) {
    const compiledQuery = db.insertInto('ecr_rr_conditions').values(ecr).compile()
  
    const {
      rows: [{ uuid }],
    } = await db.executeQuery<Pick<ECRConditions, 'uuid'>>({
      ...compiledQuery,
      sql: `${compiledQuery.sql}; select scope_identity() as uuid`,
    })
  
    return await findEcrConditionById(uuid)
}

export async function updateEcrCondition(uuid: string, updateWith: ECRConditionsUpdate) {
    await db.updateTable('ecr_rr_conditions').set(updateWith).where('uuid', '=', uuid).execute()
}

export async function deleteEcrCondition(uuid: string) {
  const ecr = await findEcrConditionById(uuid)

  if (ecr) {
    await db.deleteFrom('ecr_rr_conditions').where('uuid', '=', uuid).execute()
  }

  return ecr
}