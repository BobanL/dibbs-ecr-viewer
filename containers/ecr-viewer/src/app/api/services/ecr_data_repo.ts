import { db } from './database'
import { ECR, NewECR, ECRUpdate } from './types'

export async function findEcrById(id: string) {
  return await db.selectFrom('ecr_data')
    .where('eICR_ID', '=', id)
    .selectAll()
    .executeTakeFirst()
}

export async function findEcr(criteria: Partial<ECR>) {
  let query = db.selectFrom('ecr_data')

  if (criteria.eICR_ID) {
    query = query.where('eICR_ID', '=', criteria.eICR_ID)
  }

  if (criteria.set_id) {
    query = query.where('set_id', '=', criteria.set_id)
  }

  if (criteria.data_source) {
    query = query.where('data_source', '=', criteria.data_source)
  }

  if (criteria.fhir_reference_link) {
    query = query.where('fhir_reference_link', '=', criteria.fhir_reference_link)
  }

  if (criteria.patient_name_last !== undefined) {
    query = query.where(
      'patient_name_last',
      criteria.patient_name_last === null ? 'is' : '=',
      criteria.patient_name_last
    )
  }

  if (criteria.patient_birth_date) {
    query = query.where('patient_birth_date', '=', criteria.patient_birth_date)
  }

  if (criteria.date_created) {
    query = query.where('date_created', '=', criteria.date_created)
  }

  return await query.selectAll().execute()
}

export async function createEcr(ecr: NewECR) {
  return await db.insertInto('ecr_data')
    .values(ecr)
    .returningAll()
    .executeTakeFirstOrThrow()
}

export async function updateEcr(eICR_ID: string, updateWith: ECRUpdate) {
    await db.updateTable('ecr_data').set(updateWith).where('eICR_ID', '=', eICR_ID).execute()
}

export async function deleteEcr(eICR_ID: string) {
  const ecr = await findEcrById(eICR_ID)

  if (ecr) {
    await db.deleteFrom('ecr_data').where('eICR_ID', '=', eICR_ID).execute()
  }

  return ecr
}