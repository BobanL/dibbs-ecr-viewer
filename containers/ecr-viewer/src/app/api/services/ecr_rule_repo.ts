import { db } from './database'
import { ECRRuleSummaries, NewECRRuleSummaries, ECRRuleSummariesUpdate } from './types'

export async function findEcrRuleById(id: string) {
  return await db.selectFrom('ecr_rr_rule_summaries')
    .where('uuid', '=', id)
    .selectAll()
    .executeTakeFirst()
}

export async function findEcrRule(criteria: Partial<ECRRuleSummaries>) {
  let query = db.selectFrom('ecr_rr_rule_summaries')

  if (criteria.uuid) {
    query = query.where('uuid', '=', criteria.uuid)
  }

  if (criteria.ecr_rr_conditions_id) {
    query = query.where('ecr_rr_conditions_id', '=', criteria.ecr_rr_conditions_id)
  }

  if (criteria.rule_summary) {
    query = query.where('rule_summary', '=', criteria.rule_summary)
  }

  return await query.selectAll().execute()
}

export async function createEcrRule(ecr: NewECRRuleSummaries) {
    const compiledQuery = db.insertInto('ecr_rr_rule_summaries').values(ecr).compile()

    const {
      rows: [{ uuid }],
    } = await db.executeQuery<Pick<ECRRuleSummaries, 'uuid'>>({
      ...compiledQuery,
      sql: `${compiledQuery.sql}; select scope_identity() as uuid`,
    })

    return await findEcrRuleById(uuid)
}

export async function updateEcrRule(uuid: string, updateWith: ECRRuleSummariesUpdate) {
    await db.updateTable('ecr_rr_conditions').set(updateWith).where('uuid', '=', uuid).execute()
}

export async function deleteEcrRule(uuid: string) {
  const ecr = await findEcrRuleById(uuid)

  if (ecr) {
    await db.deleteFrom('ecr_rr_rule_summaries').where('uuid', '=', uuid).execute()
  }

   return ecr
}