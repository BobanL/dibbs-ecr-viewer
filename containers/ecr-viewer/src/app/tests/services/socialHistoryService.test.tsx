import { render } from "@testing-library/react";
import { Bundle } from "fhir/r4";

import BundleWithTravelHistory from "../../../../../../test-data/fhir/BundleTravelHistory.json";
import { evaluateTravelHistoryTable } from "@/app/services/socialHistoryService";

describe("Travel History", () => {
  it("should display a table ", () => {
    const { container } = render(
      evaluateTravelHistoryTable(BundleWithTravelHistory as unknown as Bundle),
    );
    expect(container).toMatchSnapshot();
  });
  it("should display nothing when no travel history is available", () => {
    expect(evaluateTravelHistoryTable({} as Bundle)).toBeUndefined();
  });
});
