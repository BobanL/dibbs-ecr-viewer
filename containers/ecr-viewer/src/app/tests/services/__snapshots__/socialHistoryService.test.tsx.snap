// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Travel History should display a table  1`] = `<div />`;
